features/mbedtls/targets/TARGET_NUVOTON/TARGET_M480/aes/aes_alt.c

```diff
@@ -21,298 +21,215 @@
  *  http://csrc.nist.gov/publications/fips/fips197/fips-197.pdf
  */
 
-#if !defined(MBEDTLS_CONFIG_FILE)
-#include "mbedtls/config.h"
-#else
-#include MBEDTLS_CONFIG_FILE
-#endif
+#include "mbedtls/aes.h"
 
 #if defined(MBEDTLS_AES_C)
 #if defined(MBEDTLS_AES_ALT)
 
 #include <string.h>
-
-#include "mbedtls/aes.h"
+#include <stdbool.h>
 
 #include "M480.h"
 #include "mbed_toolchain.h"
 #include "mbed_assert.h"
+#include "mbed_error.h"
+#include "nu_bitutil.h"
 
 
 
-#define mbedtls_trace(...) //printf(__VA_ARGS__)
-
 /* Implementation that should never be optimized out by the compiler */
-static void mbedtls_zeroize( void *v, size_t n ) {
-    volatile unsigned char *p = (unsigned char*)v; while( n-- ) *p++ = 0;
+static void mbedtls_zeroize( void *v, size_t n )
+{
+    volatile unsigned char *p = (unsigned char*)v;
+    while( n-- ) *p++ = 0;
 }
 
-
-static uint32_t au32MyAESIV[4] = {
-    0x00000000, 0x00000000, 0x00000000, 0x00000000
-};
-
 extern volatile int  g_AES_done;
 
-// Must be a multiple of 16 bytes block size 
+// Must be a multiple of 16 bytes block size
 #define MAX_DMA_CHAIN_SIZE (16*6)
-MBED_ALIGN(4) static uint8_t au8OutputData[MAX_DMA_CHAIN_SIZE];
-MBED_ALIGN(4) static uint8_t au8InputData[MAX_DMA_CHAIN_SIZE];
-
-static void dumpHex(const unsigned char au8Data[], int len)
-{
-		int j;									
-		for (j = 0; j < len; j++) mbedtls_trace("%02x ", au8Data[j]);									
-		mbedtls_trace("\r\n");										
-}	
 
-static void swapInitVector(unsigned char iv[16])
-{
-	  unsigned int* piv;
-	  int i;
-		// iv SWAP
-		piv = (unsigned int*)iv;
-		for( i=0; i< 4; i++)
-		{
-				*piv = (((*piv) & 0x000000FF) << 24) |
-				(((*piv) & 0x0000FF00) << 8) |
-				(((*piv) & 0x00FF0000) >> 8) |
-				(((*piv) & 0xFF000000) >> 24);
-				piv++;
-		}			
-}	
-
-/* IRQHandler: To share CRYPTO_IRQHandler() with TRNG & other crypto IPs 
-               For ex: 
-                    volatile void CRYPTO_IRQHandler()
-                    {
-                        ...
-                        if (AES_GET_INT_FLAG()) {
-                            g_AES_done = 1;
-                            AES_CLR_INT_FLAG();
-                        }
-                        ...
-                    }               
-*/
-
-
-/* AES available channel 0~3 */
-static unsigned char channel_flag[4]={0x00,0x00,0x00,0x00};  // 0: idle, 1: busy
-static int channel_alloc()
-{
-	int i;
-	for(i=0; i< (int)sizeof(channel_flag); i++)
-	{
-		if( channel_flag[i] == 0x00 )
-		{
-			channel_flag[i] = 0x01;
-			return i;
-		}	
-	}
-	return(-1);
-}
-
-static void channel_free(int i)
+/* Check if buffer can be used for AES DMA. It requires to be:
+ *   1) Word-aligned
+ *   2) Located in 0x2xxxxxxx region
+ */
+static bool aes_dma_buff_compat(const void *buff)
 {
-	if( i >=0 && i < (int)sizeof(channel_flag) )
-        channel_flag[i] = 0x00;
+    uint32_t buff_ = (uint32_t) buff;
+    
+    return (((buff_ & 0x03) == 0) && ((buff_ & 0x20000000) == 0x20000000));
```

Comment:
I am not sure this check corresponds to the conditions that you have stated in the comment above. For example consider the value `0x30000000` that is word aligned and will also satisfy the check `(buff_ & 0x20000000) == 0x20000000` even though is not in the `0x2xxxxxxx` range. Perhaps I misunderstood the constraints, but I believe something like this will satisfy:
```
ptr % 4 == 0 && (ptr & 0xF0000000) == 0x20000000
```

Reply:
@andresag01 Yes, you are right. I fixed it in https://github.com/ARMmbed/mbed-os/pull/4925/commits/3119935b1ac0be7f5f406bd49f2913ac1a1df9d5.